### PR TITLE
Alerter

### PR DESCRIPTION
**Pull request by itsbriany**
## TL;DR
Various alerts can now be sent while monitoring RabbitMQ message queues.

## Details
The message queues will be monitored based off the configuration `MONITOR_QUEUE_PREFIX`. For example, all queues whose names start with `taskcluster/` will be monitored every `MONITOR_REFRESH_INTERVAL` (in milliseconds). More specifically, a GET request is sent out to get statistics on the target message queues every `MONITOR_REFRESH_INTERVAL` milliseconds.

We only want to send alerts for queues who exceed various thresholds. Therefore, we can set these tolerances under `config.alerter`. So far, we only monitor for the message count and message upload rate exceeding a configurable threshold as shown in `config.yml`.

Finally, the type of alert is based on a namespace response. For example:

We monitor a message queue whose name is the following: `taskcluster/one`.
The namespace will be `one`.

The table holding the namespace information is then queried using the namespace `one` as a key. This allows the Rabbit Alerter to tailor its alert based off the contact information found in the namespace response.

### CODE REVIEW

File: src/rabbitalerter.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+/**
+ * An interface used to generate various TaskCluster Notifications.
+ *
+ * @class RabbitAlerter
+ */
+class RabbitAlerter {
+  /**
+   * @param {number} config.alerter.messageCountTolerance       - When a queue holds more messages
+   *                                                              than this tolerance, an alert will be created.
+   * @param {number} config.alerter.messagePublishRateTolerance - When the number of messages published per second
```
Comment (jonasfj) [inline src/rabbitalerter.js]: I suspect these are the wrong metrics... to alert on..

I think we want to alert when:
 A) **the consumer can't keep up with the number of messages published**.
 B) A non-empty queue haven't had messages consumed for a significant amount of time, say 8 minutes.

Ignoring (A) will eventually cause a full system failure (out of disk/memory).
Where as (B) is just an indicator that something crashed, for example a low through-put queue whose consumer crashed.

A fixed upper bound solves the problem for (A), until you have a queue with a really high through-put, then it might spike the fixed limit and trigger an alert. Further more the higher the fixed limit is the worse this detection is for case (B).

It's not obvious to me how to do this... Because we're only observing queue stats for points in time... We can't tell how many messages was consumed since last we looked at the stats.
Technically we have: https://gist.github.com/jonasfj/925b8652b015ff857c629cef6106fac7
(that's from a high through-out queue).

My initial thinking was that if observed that queue at every 1 min interval, if for 30 consecutive intervals there never is zero messages in the queue, then it's not being consumed.
But my strategy certainly has other downsides, such as it might not work for high-through-put queues.

Maybe the strategy is to say:
Let x[i] be the number of messages in the queue at interval i, and let y be the median of x[j], x[j+1], ... x[j+k],
if y < min(x[j+k], x[j+k-1], x[j+k-c]) then we raise an alert, where j + k is the current interval and c < k.

Essentially saying if for some c number of intervals, the message count haven't been lower than the median message count for the last k number of intervals, then we raise an alert. For this k might have to be large enough to quantify over the past week, and c might be low enough to only consider intervals from the past 30 minutes or so...

Maybe average is easier than median... who knows... Maybe there is a better option here... This was just a random thought. But this is the hard part, how do we detect that message count is trending upwards overtime, if we only observe the message count at certain snapshots.

Whatever strategy we come up with, two alerts are probably safe to do:
 1) Queue has reach some obscenely high number of messages (say 1M)
 2) Queue haven't seen any consumption in say 25 minutes or so. 

While those are definitely sane, (1) won't catch crashed consumers soon enough, and never catch low-through-put queues. (2) won't do anything high-through-put queues that consumers can't keep up with.

------

Furthermore, I suspect we want to alert people if a queue is non-empty and nothing has been consumed for a significant amount of time, say 8 minutes.

Small queues with only a few messages that never gets consumed or deleted aren't going to cause system failures.
But they are indicators that something broke somewhere, so it's nice of us to alert people about that.

Comment (jonasfj) [inline src/rabbitalerter.js]: sorry, for the ramble... I know this is the hard part.

Comment (itsbriany) [inline src/rabbitalerter.js]: Thanks for the clarification @jonasfj!

The message below is just to verify my understanding.
Let me know if I am missing anything, or if what I say is inaccurate.

In a nutshell, when the publishing rate of messages to a queue is higher than that queue's consumption rate, then we send an alert. (If the publish rate is higher than the consumption rate, eventually, the system will fail due to too many messages in the queue). 

We also do not want to send too many false positive alerts. For example, a queue has zero consumption rate, and one message is being published to it a day. However, since the publish rate is higher than the consumption rate, alerts would be sent either way.

I think a solution for this would be to set a threshold tolerance for alerting where we take the following:
**message publish rate / message consumption rate**. 

Once this tolerance is exceeded, then we can send an alert. Of course, there are exceptions where there are intervals where the message consumption rate will be zero. 

When the consumption rate is zero and the publish rate is greater than zero, when we wait 8 minutes. If no messages have been consumed, then this queue is deleted.

Comment (jonasfj) [inline src/rabbitalerter.js]: > when the publishing rate of messages to a queue is higher than that queue's consumption rate

Yeah, but that is easier said than done... Because "consumption rate" will never be higher than "publishing rate". And "publishing rate" should definitely be allowed to spike :)

> We also do not want to send too many false positive alerts. For example, a queue has zero consumption rate, and one message is being published to it a day.

Yeah, probably we want some threshold to say if queue has less than 100 messages we never alert... Unless it's been like that for a day..

> We can see how many messages were consumed by keeping track of the previous queue snapshot (each snapshot will show how many messages the queue has and when). We know that messages were consumed when the current snapshot shows less messages than the previous one.

Well, messages are both published and consumed in-between snapshots, so total number of messages in the queue is of little use...

If you look at: https://gist.github.com/jonasfj/925b8652b015ff857c629cef6106fac7
Then perhaps `message_stats` can be used to figure out number of messages published and consumed between two snapshots...

Last @imbstack and I talked about it, our thinking was that what we really wanted was to know how old the oldest message in the queue was and alert if it was older than say 15 minutes. Maybe even delete the queue if older than 7 days and nobody reacted on it...

So just brainstorming here... but what if we stored the `message_stats.publish` and `message_stats.ack` (acknowledged messages are those fully processed) for each minute for the past say 30 minutes or so?
Wouldn't we then get something like:
```
time: 13:00  published:   0  acked:    0  queue:  0  oldest: 13.00  max-age: 0 min
time: 13:01  published:  10  acked:    5  queue:  5  oldest: 13.01  max-age: 0 min
time: 13:02  published:  50  acked:   20  queue: 30  oldest: 13.02  max-age: 0 min
time: 13:03  published:  50  acked:   30  queue: 20  oldest: 13.02  max-age: 1 min
time: 13:04  published:  50  acked:   35  queue: 15  oldest: 13.02  max-age: 2 min
time: 13:05  published:  50  acked:   40  queue: 10  oldest: 13.02  max-age: 3 min
time: 13:06  published:  50  acked:   45  queue:  5  oldest: 13.02  max-age: 4 min
time: 13:07  published:  60  acked:   50  queue: 10  oldest: 13.07  max-age: 0 min
time: 13:08  published:  60  acked:   60  queue:  0  oldest: 13.08  max-age: 0 min
```

Wouldn't we then be able to say at `13:06` the oldest message in the queue was ~4 minutes old.
Similarly, couldn't we store the published and acked numbers for a snapshot every day... Enabling us to see if all messages published yesterday have been consumed.

Let's say we keep:
 A) a `{published: ..., acked: ...}` snapshot for each minute for the last 30 minutes.
 B) a `{published: ..., acked: ...}` for the start of each day for the past 7 days.

Then we can alarm if:
  * All messages older than 30 min haven't been consumed (using (A))  
  * All messages older than 3 days haven't been consumed (using (B))
  * All messages older than 4 days haven't been consumed (using (B))
  * All messages older than 5 days haven't been consumed (using (B))
  * All messages older than 6 days haven't been consumed (using (B))

And finally, we can delete the queue if:
 1. all messages older than 7 days haven't been consumed, and
 2. the queue contains more than 1000 messages.

I suggest adding condition (2) because there is no real value in deleting queues with less than 1000 messages. They aren't going to cause us problems.

Maybe we need an additional delete scenario like: delete queue if:
 1. all messages older than 2 days haven't been consumed, and
 2. the queue contains more than 100k messages.
Based on the assumption that 100k messages might start to eat up system resources.

Comment (jonasfj) [inline src/rabbitalerter.js]: Note: This assumes that `message_stats.publish` is actually the absolute number of messages published since the queue was created. I hope that is the case, and I hope that is present for all queues, we should have a look, maybe it's only high traffic queues that has this... Or maybe it's only queues with active consumers.
(I guess if it requires active consumers we can alert if `message_stats` it's not present 30 min).

Also we would have to store the absolute, number so that if the number goes down (because the queue was deleted and recreated) then we just discard whatever history we have and start over.

Comment (Hexcles) [inline src/rabbitalerter.js]: Sorry for bumping this up so late. I was struggling in the swamps of deadlines and exams last week...

@jonasfj 's idea of inferring max age (oldest messages) from `message_stats.publish` and `message_stats.ack` sounds awesome to me, provided that these two counters are reliable (and are what we think they are - monotonic counters of the whole history since the server starts). In the [latest doc of rabbitmq-management](https://raw.githack.com/rabbitmq/rabbitmq-management/rabbitmq_v3_6_6/priv/www/doc/stats.html), `publish` is " Count of messages published. ", which doesn't really help much. Besides, there is no `ack` in the latest doc. It might have been renamed in recent versions, or is undocumented. I do notice that rabbitmq introduced breaking changes to the management API sometimes, and we need to check the doc of the version on the production servers to be sure. (As a side note, it is hence somewhat hard to make a reusable package for the rabbitmq management API because of the version dependency.)

FWIW, I just checked out how [Sensu](https://sensuapp.org/), a monitoring system I used before, monitors rabbitmq. One metric they chose is the (estimated) "draining time":
`messages` / `backing_queue_status.avg_egress_rate` ([code](https://github.com/sensu-plugins/sensu-plugins-rabbitmq/blob/master/bin/check-rabbitmq-queue-drain-time.rb)).

Again, it is not clear to me in what time period the rate is measured and how it is averaged (I'm guessing it measures rates at a short sampling interval and averages the rates over a few intervals). If the `avg_egress_rate` is averaged over a long period, we might give false alerts after "processable" spikes (a spike in a low-traffic queue but consumers are able to crank up).

Anyway, this is a good starting point as we have the basic facilities for monitoring and alerting now. Nice work @itsbriany !
